File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/TopInterview1Round/mergeKLists_23.java
package algo.tzashinorpu.SpecialSubject.TopInterview1Round;

import algo.tzashinorpu.FourthRound.Chapter03.ListNode;

import java.util.Comparator;
import java.util.PriorityQueue;

public class mergeKLists_23 {


    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;
        PriorityQueue<ListNode> queue = new PriorityQueue<>(lists.length, new Comparator<ListNode>() {
            @Override
            public int compare(ListNode o1, ListNode o2) {
                if (o1.val < o2.val) return -1;
                else if (o1.val == o2.val) return 0;
                else return 1;
            }
        });
        ListNode dummy = new ListNode(0);
        ListNode p = dummy;
        for (ListNode l : lists) {
            if (l != null) queue.add(l);
        }
        while (!queue.isEmpty()) {
            p.next = queue.poll();
            p = p.next;
            if (p.next != null) queue.add(p.next);
        }
        return dummy.next;
    }

    public ListNode mergeKListsDivide(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;
        return merge(lists, 0, lists.length - 1);
    }

    private ListNode merge(ListNode[] lists, int l, int r) {
        if (l == r) return lists[l];
        if (l > r) return null;
        int mid = (l + r) >> 1;
        return merge2Lists(merge(lists, l, mid), merge(lists, mid + 1, r));
    }

    public ListNode merge2Lists(ListNode l1, ListNode l2) {
        ListNode dummyNode = new ListNode(0);
        ListNode p = dummyNode;
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                p.next = l1;
                l1 = l1.next;
            } else {
                p.next = l2;
                l2 = l2.next;
            }
            p = p.next;
        }
        if (l1 != null) p.next = l1;
        if (l2 != null) p.next = l2;
        return dummyNode.next;
    }
}

File: Clion/ZeroTrac/array_list/mergeKLists_23.cpp
//
// Created by TzashiNorpu on 10/17/2022.
//
#include "array_list.h"

using namespace Array_List;


ListNode *Solution::mergeKLists(vector<ListNode *> &lists) {
  struct compare {
      bool operator()(const ListNode *l, const ListNode *r) {
        return l->val > r->val;
      }
  };
  ListNode *dummy = new ListNode();
  ListNode *p = dummy;
  priority_queue<ListNode *, vector<ListNode *>, compare> q;
  for (auto &list: lists)
    if (list)q.push(list);
  while (!q.empty()) {
    ListNode *top = q.top();
    p->next = top;
    if (q.top()->next)
      q.push(q.top()->next);
    q.pop();
    p = p->next;
  }
  return dummy->next;
}